### PR TITLE
Upgrade to the WASM toolchain

### PR DESCRIPTION
**Pull request by sokra**
**What kind of change does this PR introduce?**
refactoring, bugfixes

**Did you add tests for your changes?**
yes

**If relevant, link to documentation update:**
N/A

**Summary**

**Does this PR introduce a breaking change?**
yes to experimental features (WASM)

**Other information**

### CODE REVIEW

File: lib/wasm/WasmMainTemplatePlugin.js [hunk 1]
```diff
@@ -66,7 +66,7 @@ function generateImportObject(module) {
 						`[${JSON.stringify(data.exportName)}]`
 				);
 			} else {
-				const params = data.description.params.map(
+				const params = data.description.signature.params.map(
```
Comment (xtuc) [inline lib/wasm/WasmMainTemplatePlugin.js]: good catch! sorry for having missed that

File: declarations.d.ts [hunk 2]
```diff
@@ -29,6 +29,66 @@ declare module "chrome-trace-event" {
 	}
 }
 
+// There are no typings for @webassemblyjs/ast
+declare module "@webassemblyjs/ast" {
```
Comment (xtuc) [inline declarations.d.ts]: Could I move these declaration into the `@webassembly` namespace?

Comment (sokra) [reply declarations.d.ts]: Currently these are the bare minimum types required for compile successful. If you want to move this to your monorepo or DT you probably need full typings.

File: lib/wasm/WebAssemblyGenerator.js [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+
+	if (typeof startAtFuncIndex !== "undefined") {
+		funcBody.push(t.callInstruction(startAtFuncIndex));
+	}
+
+	const funcResults = [];
+
+	const func = t.func(initFuncId, funcParams, funcResults, funcBody);
+
+	const functype = t.typeInstructionFunc(func.params, func.result);
```
Comment (xtuc) [inline lib/wasm/WebAssemblyGenerator.js]: We also need to add `signature` here

```js
const functype = t.typeInstructionFunc(func.signature.params, func.signature.results);
```

File: lib/wasm/WasmMainTemplatePlugin.js (hunk 1 above)
Comment (sokra) [reply lib/wasm/WasmMainTemplatePlugin.js]: It's a bit weird to have this kind of breaking change on a patch version...

Comment (xtuc) [reply lib/wasm/WasmMainTemplatePlugin.js]: I agree, that was my bad

File: lib/wasm/WebAssemblyGenerator.js (hunk 3 above)
Comment (sokra) [reply lib/wasm/WebAssemblyGenerator.js]: done

**Comment by sokra** [comment]
Yeah

**Comment by sokra** [comment]
Thanks

**Pull request merged**